src/common/atomic_queue.cpp

```diff
@@ -0,0 +1,397 @@
+    std::map <fi_addr_t,fi_addr_t> fiAddrMap;
+    std::map<fi_addr_t,fi_addr_t>::iterator it;
+
+    // Recover the incomplete writes if any
+    cout << "Recovering Incomplete transactions for queue" << lcTInfo->qId << endl;
+    ret = recover_queue(lcTInfo->qId,allocator);
+    cout << "Recovery of Incomplete trasactions completed for queue" << lcTInfo->qId << endl;
+    cout << "processing thread for queue" << lcTInfo->qId << " started" << endl;
+
+    while (1) {
+        ret = atomicQ[lcTInfo->qId].read(&item);
+        if (ret == ATOMIC_QUEUE_EMPTY) {//queue empty
+            pthread_mutex_lock(&mutex[lcTInfo->qId]);
+            pthread_cond_wait(&empty[lcTInfo->qId], &mutex[lcTInfo->qId]);
+            pthread_mutex_unlock(&mutex[lcTInfo->qId]);
+            continue;
+        }
+        // get fiAddr of the remote/client node
+        localPointer = allocator->get_local_pointer(item.regionId,item.offset);
+        msgPointer = (atomicMsg *)localPointer;
+        remoteAddr = (char *)calloc(1,msgPointer->nodeAddrSize);
+        memcpy(remoteAddr,msgPointer->nodeAddr,msgPointer->nodeAddrSize);
+        clientAddr = *(fi_addr_t *)remoteAddr;
+        it = fiAddrMap.find(clientAddr);
+        if (it == fiAddrMap.end()) {
+            fiAddr = fabric_insert_av_atomic(remoteAddr,famOpsLibfabricQ->get_av());
+            fiAddrMap.insert (it, std::pair<fi_addr_t,fi_addr_t>(clientAddr,fiAddr) );
+        }
+        else fiAddr = it->second;
+        retStatus = 0;
+        uint64_t function = msgPointer->flag & 0x3F;
+        switch (function) {
+            case ATOMIC_READ : {//Read function
+                try {
+                    // Get the pointer for the source using the offset
+                    localPointerD = allocator->get_local_pointer(msgPointer->dstDataGdesc.regionId,
+				msgPointer->dstDataGdesc.offset+msgPointer->offset);
+                    // Write data to client's memory
+                    ret = fabric_write(msgPointer->key,localPointerD,msgPointer->size,0, fiAddr, 
+				famOpsLibfabricQ->get_defaultCtx(uint64_t(0)));
+                } catch (Memory_Service_Exception &e) {
+                    cout << "Error code: " << e.fam_error() << endl;
+                    cout << "Error msg: " << e.fam_error_msg() << endl;
+                    retStatus = DATAITEMNOTFOUND;
+                }
+
+                // send status (retStatus)  back to client
+                fabric_send_response(&retStatus,fiAddr,famOpsLibfabricQ->get_defaultCtx(uint64_t(0)),sizeof(retStatus));
+                break;
+            }
+            case ATOMIC_WRITE : {
+                // Atomic write
+                if (msgPointer->flag & ATOMIC_WRITE_COMPLETED) {
+                    //last write completed. If buffer is still allocated, deallocate and  remove the entry
+                    if (msgPointer->flag & ATOMIC_BUFFER_ALLOCATED)
+                        allocator->deallocate(ATOMIC_REGION_ID,msgPointer->offsetBuffer);
```

Comment:
exception wherever we call any functions which can throw exception. process_queue threads should never crash.